### ISSUE TITLE
Stats APIs for C++

### ISSUE THREAD

**Issue by alyssawilk** [envoyproxy/envoy-mobile#2356]
While doing https://github.com/envoyproxy/envoy-mobile/issues/2303 we uncovered a bunch of gaps between the C APIs and C++.
I'm fixing those but as discussed at this week's meeting we're going to punt on stats and leave pulseClient a TODO so filing as API tech debt bug.

**Comment by abeyad** [comment]
Just for clarification, there are two different concepts here in the Envoy Mobile world:

(1) The `PulseClient` is for creating and updating stats (counters, gauges, timers, distributions) for the mobile client only.  The stats generated by the `PulseClient` use a `mobile.` prefix.  They are specifically meant to create stats on the mobile client, not for interacting with Envoy Engine/Server stats.

(2) An ability to get stats (counters, etc) from the Envoy Engine.  The primary use case for this is integration tests.  Otherwise, to export stats, one must use a stats sink.

Previously, the only way to do (2) was to use the admin interface.  However, we recently added the ability to get counters and wait on counter values in integration tests: https://github.com/envoyproxy/envoy/blob/main/mobile/test/common/integration/base_client_integration_test.h#L65-L68.  Hence, we don't need the admin interface for (2) and can run the mobile integration tests with the admin interface compiled out.  So our needs for (2) are taken care of.

For (1), the PulseClient probably needs some rethinking.  In discussing with @Augustyniak, the only PulseClient API used and worth keeping are for counters.  Hence, we got rid of the gauges, timers, and distribution APIs in https://github.com/envoyproxy/envoy/pull/24648 and https://github.com/envoyproxy/envoy/pull/24645.  However, there are more inconsistencies with the PulseClient:
  - It only exists for Kotlin and Swift.  The Java and Obj-C APIs do not have a PulseClient.  Instead, we use functions like: https://github.com/envoyproxy/envoy/blob/357bf0016ffa0102ed1fc3c51687eb23b9d2305d/mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyEngine.java#L47-L105
  - C++ APIs do not have any PulseClient support at all, not through PulseClient nor through Engine APIs like Java and Obj-C.

So the next steps for this bug are:
  - Remove the "recordGauge", "recordTimer", "recordHistogram" functions from the Java and Obj-C APIs, keeping only the counter related functions.
  - Move the Java and Obj-C APIs "recordCounterInc" API to a new PulseClient class, so we have consistent APIs between Java, Obj-C, Swift, and Kotlin.
  - Add a PulseClient for the C++ APIs.  The class already exists, but it does nothing: https://github.com/envoyproxy/envoy/blob/main/mobile/library/cc/pulse_client.h.  All we need to do is add the `counter` function to the C++ PulseClient.

**Comment by abeyad** [comment]
Note: since we have some other higher priorities and we've unblocked getting counters/gauges in integration tests without relying on the admin interface, I (abeyad) am putting work on this on hold for the time being.

**Comment by Augustyniak** [comment]
> * It only exists for Kotlin and Swift.  The Java and Obj-C APIs do not have a PulseClient.  

I think that the idea is that that you can use existing Swift APIs from within objective-c and existing Kotlin APIs from within Java. That's why we do not need to have separate objective-c and Java implementations. That being said, we do miss c++ Pulse APIs (as you have mentioned).

>Instead, we use functions like: https://github.com/envoyproxy/envoy/blob/357bf0016ffa0102ed1fc3c51687eb23b9d2305d/mobile/library/java/io/envoyproxy/envoymobile/engine/EnvoyEngine.java#L47-L105

The referenced methods are not supposed to be called by external clients (from outside of the Envoy Mobile library). They are used by the implementation of Kotlin Pulse APIs to move from Kotlin context to shared c++ context:

https://github.com/envoyproxy/envoy/blob/f99dd64a016a92180bb74b33c0e4940ca190bb08/mobile/library/common/engine.cc#L186-L196

**Comment by abeyad** [comment]
> I think that the idea is that that you can use existing Swift APIs from within objective-c and existing Kotlin APIs from within Java. That's why we do not need to have separate objective-c and Java implementations. 

Is that the convention we use for all APIs (i.e. just create Swift/Kotlin ones and the Java/Obj-C implementations can just invoke the Swift/Kotlin ones)?
